Make API compatible with SQL backend

NoahCarnahan opened the pull request:
@dimagi/scale-team cc @snopoke 

Best reviewed commit-by-commit

NoahCarnahan committed:
Limit api test dependencies slightly

NoahCarnahan committed:
Test xfrom resource cases field

NoahCarnahan committed:
Make xform api db agnostic

NoahCarnahan committed:
Move shared methods into abstract form class

NoahCarnahan committed:
Make case api backend agnostic

NoahCarnahan committed:
Test case resource parent and child cases fields

NoahCarnahan committed:
Fix case resource parent_cases and child_cases

NoahCarnahan committed:
Move shared case methods into abstract case class

dimagimon commented:
There are 5 errors:

* corehq/apps/api/resources/v0_1.py, line 10 - F401 'XFormNotFound' imported but unused
* corehq/apps/api/resources/v0_1.py, line 11 - F401 'FormAccessors' imported but unused
* corehq/apps/api/resources/v0_1.py, line 43 - F401 'object_does_not_exist' imported but unused
* corehq/apps/api/resources/v0_4.py, line 217 - E302 expected 2 blank lines, found 1
* corehq/form_processor/models.py, line 576 - E303 too many blank lines (2)

NoahCarnahan commented on corehq/apps/api/resources/v0_4.py:
```diff
@@ -216,10 +216,10 @@ class CommCareCaseResource(SimpleSortableResourceMixin, v0_3.CommCareCaseResourc
 
     parent_cases = UseIfRequested(ToManyDictField('corehq.apps.api.resources.v0_4.CommCareCaseResource',
-                                                  attribute=lambda case: dict([ (index.identifier, CommCareCase.get(index.referenced_id)) for index in case.reverse_indices])))
+                                                  attribute=lambda case: dict([ (index.identifier, CommCareCase.get(index.referenced_id)) for index in case.indices])))
```
Apparently the parent_cases field was returning the child cases (and vice versa) for some time. These fields don't seem to be documented on the confluence page though (and you need to add `?parent_cases__full=true` to the url to get it), so hopefully no one has been relying on the erroneous behavior.

benrudolph replied:
haha :flushed:

snopoke replied:
Should we add docs?

NoahCarnahan replied:
Yeah, I'll update the confluence page once this is deployed.

NoahCarnahan committed:
lint

benrudolph commented on corehq/apps/api/resources/v0_3.py:
```diff
@@ -68,22 +69,11 @@ class XFormInstanceResource(v0_1.XFormInstanceResource, DomainSpecificResourceMi
-        except (ResourceNotFound, couchforms.UnexpectedDeletedXForm):
-            pass  # covered by the below
-        except AttributeError:
-            # there's a weird edge case if you reference a form with a case id
```
lol

benrudolph commented on corehq/ex-submodules/casexml/apps/case/xform.py:
```diff
@@ -366,16 +366,14 @@ def get_case_ids_from_form(xform):
-            assert case.domain == domain
-
-    return cases
+    case_accessor = CaseAccessors(domain)
```
did you find instances when we don't have the domain? seems bad

NoahCarnahan replied:
No I didn't find instances of that happening. This function seems like it had been written to account for instances when domain is `None`, so thought I would preserve that behavior.

benrudolph commented on corehq/form_processor/abstract_models.py:
```diff
@@ -155,6 +155,65 @@ def get_subcases(self):
     def get_case_property(self, property):
         raise NotImplementedError
 
+    def get_json(self, lite=False):
```
we should use our serializers for this if possible: https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/serializers.py

NoahCarnahan replied:
Good point.
After reviewing this comment I realized it was actually overkill to move this whole function to `AbstractCommCareCase`, so I'm going to move it back to `CommCareCase`, and just use tastypie's serialization workflow to grab the bits it needs (properties and indices) while leaving `get_json()` unchanged: https://github.com/dimagi/commcare-hq/pull/10976/commits/e84e725d6241b167966d2262c3e2d96004a6fbde

benrudolph replied:
:+1:

benrudolph commented:
:candy:

NoahCarnahan committed:
Move get_json back out of AbstractCommCareCase; don't use get_json in CommCareCaseResource

benrudolph commented:
test failure real:
```
======================================================================
FAIL: testCaseAPIResultJSON (corehq.apps.cloudcare.tests.test_api.CaseAPITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/apps/cloudcare/tests/test_api.py", line 233, in testCaseAPIResultJSON
    self.assertEqual(json['properties']['case_type'], None)
AssertionError: '' != None
```

NoahCarnahan committed:
Fix cloudcare test

CaseAPIResult.case_json() modified mutated the value returned by
AbstractCommCareCase.get_properties_in_api_format(), and since
this function was memoized, the changes persisted into subsequent
calls. This commit fixes it by both not memoizing that function
and not mutating the result.

NoahCarnahan commented:
:white_check_mark:
